setools/diff/terules.py

```diff
@@ -38,32 +41,138 @@
 modified_terule_record = namedtuple("modified_terule", ["rule", "added_default", "removed_default"])
 
 
-def _avrule_expand_generator(rule_list, WrapperClass):
+def _avrule_expand_generator(rule_list, rule_db, type_db, side):
     """
-    Generator that yields wrapped, expanded, av(x) rules with
-    unioned permission sets.
+    Using rule_list, build up rule_db which is a data structure which consists
+    of nested dicts that store BOTH the left and the right policies. All of the
+    keys are interned strings. The permissions are stored as a set. The basic
+    structure is rule_db[cond_exp][block_bool][src][tgt][tclass][side]=perms
+    where:
+      cond_exp is a boolean expression
+      block_bool is either true or false
+      src is the source type
+      tgt is the target type
+      tclass is the target class
+      side is either left or right
+      perms is the set of permissions for this rule
+    There are a few advantages to this structure. First, it takes up way less
+    memory. Second, it allows redundant rules to be easily eliminated. And,
+    third, it makes it easy to create the added, removed, and modified rules.
     """
-    items = dict()
+    if side not in type_db:
+        type_db[side] = dict()
 
     for unexpanded_rule in rule_list:
-        for expanded_rule in unexpanded_rule.expand():
-            expanded_wrapped_rule = WrapperClass(expanded_rule)
-
-            # create a hash table (dict) with the first rule
-            # as the key and value.  Rules where permission sets should
-            # be unioned together have the same hash, so this will union
-            # the permissions together.
-            try:
-                items[expanded_wrapped_rule].perms |= expanded_wrapped_rule.perms
-            except KeyError:
-                items[expanded_wrapped_rule] = expanded_wrapped_rule
-
-    if items:
-        logging.getLogger(__name__).debug(
-            "Expanded {0.ruletype} rules for {0.policy}: {1}".format(
-                unexpanded_rule, len(items)))
-
-    return items.keys()
+        cond_exp = terules_unconditional
+        block_bool = terules_unconditional_block
+        try:
+            cond_exp = intern(str(unexpanded_rule.conditional))
+            block_bool = intern(str(unexpanded_rule.conditional_block))
+        except RuleNotConditional:
+            pass
+
+        if cond_exp not in rule_db:
+            rule_db[cond_exp] = dict()
+            rule_db[cond_exp][block_bool] = dict()
+        elif block_bool not in rule_db[cond_exp]:
+            rule_db[cond_exp][block_bool] = dict()
+
+        tclass = intern(str(unexpanded_rule.tclass))
+        perms = {intern(str(p)) for p in unexpanded_rule.perms}
+        side_data = (perms, unexpanded_rule)
```

Comment:
Using a named tuple in the new version.

Reply:
By the way I am not generating expanded rules. I keep the original rule so that an expanded rule can be created just for the rules that are added, removed, or modified. This saves a ton of memory. Also I create the tuple here on purpose so that it can be shared among the expanded rules. This also saves a ton of memory because newer Fedora policies have 69 rules that expand out to more than 725,000 rules and 4 rules that expand out to over 2,500,000 rules. If more permissions are added then a new tuple is created so the shared one is not modified.